src/pytorch_lightning/loggers/neptune.py

```diff
@@ -649,11 +650,11 @@ def _signal_deprecated_api_usage(f_name, sample_code, raise_exception=False):
             raise ValueError("The function you've used is deprecated.\n" + msg_suffix)
 
     @rank_zero_only
-    def log_metric(self, metric_name: str, metric_value: Union[Tensor, float, str], step: Optional[int] = None):
+    def log_metric(self, metric_name: str, metric_value: Union[Tensor, float, str], step: Optional[int] = None) -> None:
         key = f"{self._prefix}/{metric_name}"
         self._signal_deprecated_api_usage("log_metric", f"logger.run['{key}'].log(42)")
         if torch.is_tensor(metric_value):
-            metric_value = metric_value.cpu().detach()
+            metric_value = metric_value.cpu().detach()  # type: ignore[union-attr]
```

Comment:
types `float` and `str` cause the following error despite `if torch.is_tensor(metric_value):` checking the type before executing `metric_value.cpu().detach()`

```sh
 Item ...  of "Union[Tensor, float, str]" has no attribute "cpu"
```

Reply:
There is a check `torch.is_tensor(metric_value)` above, which means we know that it can only be a tensor. MyPy doesn't know this, but we can tell it this by 
`assert isinstance(metric_value, Tensor)`

IMO this should be done instead of the type ignore

Reply:
I think future contributors/maintainers might see the assert and wonder why it was added after the is_tensor check, then remove it; whereas ignoring the error handles for mypy checks only, and does not introduce new code just to appease a mypy error.

In general, it would probably be good to set a convention on when, why, and how to change code or add ignore[union-attr] given the amount of optionals & unions in the code base.

Reply:
Actually you are right, we should change it to 

`if isinstance(metric_value, Tensor):`

this is even better. Then the assert is not needed and the ignore can be dropped.

Reply:
updated in https://github.com/Lightning-AI/lightning/pull/13692/commits/4f049a97ba3d32ccb18b51320b85efecd67ad452